common/src/main/java/org/conscrypt/SSLUtils.java

```diff
@@ -253,28 +254,76 @@ static String getClientKeyType(byte clientCertificateType) {
+            case NativeConstants.SSL_SIGN_ECDSA_SECP521R1_SHA512:
+                // We have no way to communicate the curve to the KeyManager, so we'll just
+                // have to hope the EC key it chooses uses a supported curve.
+                return KEY_TYPE_EC;
+            default:
+                return null;
+        }
+    }
+
     /**
      * Gets the supported key types for client certificates based on the
      * {@code ClientCertificateType} values provided by the server.
      *
-     * @param clientCertificateTypes {@code ClientCertificateType} values provided by the server.
-     *        See https://www.ietf.org/assignments/tls-parameters/tls-parameters.xml.
+     * @param clientCertificateTypes
+     *         {@code ClientCertificateType} values provided by the server.
+     *         See https://www.ietf.org/assignments/tls-parameters/tls-parameters.xml#tls-parameters-2.
+     * @param signatureAlgs
+     *         {@code SignatureScheme} values provided by the server.
+     *         See https://www.ietf.org/assignments/tls-parameters/tls-parameters.xml#tls-signaturescheme
      * @return supported key types that can be used in {@code X509KeyManager.chooseClientAlias} and
-     *         {@code X509ExtendedKeyManager.chooseEngineClientAlias}.
+     * {@code X509ExtendedKeyManager.chooseEngineClientAlias}.
      *
      * Visible for testing.
      */
-    static Set<String> getSupportedClientKeyTypes(byte[] clientCertificateTypes) {
-        Set<String> result = new HashSet<String>(clientCertificateTypes.length);
+    static Set<String> getSupportedClientKeyTypes(byte[] clientCertificateTypes,
+            int[] signatureAlgs) {
+        Set<String> fromClientCerts = new HashSet<String>(clientCertificateTypes.length);
         for (byte keyTypeCode : clientCertificateTypes) {
             String keyType = SSLUtils.getClientKeyType(keyTypeCode);
             if (keyType == null) {
                 // Unsupported client key type -- ignore
                 continue;
             }
-            result.add(keyType);
+            fromClientCerts.add(keyType);
+        }
+        Set<String> fromSigAlgs = new HashSet<String>(signatureAlgs.length);
+        for (int signatureAlg : signatureAlgs) {
+            String keyType = SSLUtils.getClientKeyTypeFromSignatureAlg(signatureAlg);
+            if (keyType == null) {
+                // Unsupported client key type -- ignore
+                continue;
+            }
+            fromSigAlgs.add(keyType);
+        }
+        // If both are specified, the key needs to meet both sets of requirements.  Otherwise,
+        // just meet the set of requirements that were specified.  See RFC 5246, section 7.4.4.
+        // (In TLS 1.3, certificate_types is no longer used and is never present.)
+        if (clientCertificateTypes.length > 0 && signatureAlgs.length > 0) {
+            fromClientCerts.retainAll(fromSigAlgs);
```

Comment:
nit: Not really necessary, but spec wise, fromSigAlgs.retainAll(fromClientCerts) might be slightly more correct even though these behave identically. (This is more relevant if you used the priority ordering).